### PR TITLE
Docs: add more cross-references to the KV cache docs

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

@zucchini-nlp wrote a great guide for our KV caches. This PR adds more cross-references to that guide in relevant places. 

These references were added because: 
1. a docstring can't fully explain what a cache can be 
2. users have noticed that our cache classes are underdocumented/not easy to find (e.g. #33270)

### CODE REVIEW

File: docs/source/en/kv_cache.md [hunk 1]
```diff
@@ -51,11 +51,11 @@ More concretely, key-value cache acts as a memory bank for these generative mode
 
 
   See an example below for how to implement your own generation loop.
-    
+
```
Comment (gante) [inline docs/source/en/kv_cache.md:54]: (my VSCode automatically removes any hanging space on save 🤗 )

File: src/transformers/generation/configuration_utils.py [hunk 2]
```diff
@@ -44,6 +53,16 @@
 METADATA_FIELDS = ("_from_model_config", "_commit_hash", "_original_object_hash", "transformers_version")
 NEEDS_CACHE_CONFIG = {}
 
+NEED_SETUP_CACHE_CLASSES_MAPPING = {
```
Comment (gante) [inline src/transformers/generation/configuration_utils.py:56]: moved these constants to the file holding `GenerationConfig`, so we can validate cache config options

File: src/transformers/models/codegen/modeling_codegen.py [hunk 3]
```diff
@@ -429,7 +429,8 @@ def _init_weights(self, module):
             returned by the model at a previous stage of decoding, when `use_cache=True` or `config.use_cache=True`.
 
             Two formats are allowed:
-            - a [`~cache_utils.Cache`] instance;
+            - a [`~cache_utils.Cache`] instance, see our
+            [kv cache guide](https://huggingface.co/docs/transformers/en/kv_cache);
```
Comment (gante) [inline src/transformers/models/codegen/modeling_codegen.py:433]: Update added to all models supporting `Cache` classes

File: src/transformers/models/mistral/modeling_tf_mistral.py [hunk 4]
```diff
@@ -728,8 +728,7 @@ class TFMistralPreTrainedModel(TFPreTrainedModel):
             blocks) that can be used to speed up sequential decoding. This typically consists in the `past_key_values`
             returned by the model at a previous stage of decoding, when `use_cache=True` or `config.use_cache=True`.
 
-            Two formats are allowed:
-            - a [`~cache_utils.Cache`] instance;
```
Comment (gante) [inline src/transformers/models/mistral/modeling_tf_mistral.py:732]: This is not true for TF models 👀

**Review by LysandreJik — APPROVED** [review]
Nice, very welcome cross-references. This looks exhaustive, LGTM

File: src/transformers/generation/configuration_utils.py [hunk 5]
```diff
@@ -146,7 +169,10 @@ class GenerationConfig(PushToHubMixin):
             Whether or not the model should use the past last key/values attentions (if applicable to the model) to
             speed up decoding.
         cache_implementation (`str`, *optional*, default to `None`):
-            Cache class that should be used when generating.
+            Name of the cache class that will be instantiated in `generate`, for fasted decoding. Possible values are:
```
Comment (gante) [inline src/transformers/generation/configuration_utils.py:172]:
```suggestion
            Name of the cache class that will be instantiated in `generate`, for faster decoding. Possible values are:
```
typo

**Review by zucchini-nlp — APPROVED** [review]
Thanks so much for adding this 💖

**Pull request merged**